flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java

```diff
@@ -0,0 +1,352 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.CatalogNotExistException;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.catalog.exceptions.CatalogException;
+import org.apache.flink.table.catalog.exceptions.TableNotExistException;
+import org.apache.flink.table.factories.TableFactoryUtil;
+import org.apache.flink.table.operations.CatalogTableOperation;
+import org.apache.flink.table.sinks.TableSink;
+import org.apache.flink.util.StringUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A CatalogManager that encapsulates all available catalogs. It also implements the logic of
+ * table path resolution. Supports both new API ({@link Catalog} as well as {@link ExternalCatalog}).
+ */
+@Internal
+public class CatalogManager {
+	private static final Logger LOG = LoggerFactory.getLogger(CatalogManager.class);
+
+	// A map between names and catalogs.
+	private Map<String, Catalog> catalogs;
+
+	// TO BE REMOVED along with ExternalCatalog API
+	private Map<String, ExternalCatalog>  externalCatalogs;
+
+	// The name of the default catalog and schema
+	private String currentCatalogName;
+
+	private String currentDatabaseName;
+
+	public CatalogManager(String defaultCatalogName, Catalog defaultCatalog) {
+		checkArgument(
+			!StringUtils.isNullOrWhitespaceOnly(defaultCatalogName),
+			"Default catalog name cannot be null or empty");
+		checkNotNull(defaultCatalog, "Default catalog cannot be null");
+		catalogs = new LinkedHashMap<>();
+		externalCatalogs = new LinkedHashMap<>();
+		catalogs.put(defaultCatalogName, defaultCatalog);
+		this.currentCatalogName = defaultCatalogName;
+		this.currentDatabaseName = defaultCatalog.getDefaultDatabase();
+	}
+
+	/**
+	 * Registers a catalog under the given name. The catalog name must be unique across both
+	 * {@link Catalog}s and {@link ExternalCatalog}s.
+	 *
+	 * @param catalogName name under which to register the given catalog
+	 * @param catalog catalog to register
+	 * @throws CatalogException if the registration of the catalog under the given name failed
+	 */
+	public void registerCatalog(String catalogName, Catalog catalog) {
```

Comment:
I might miss it somewhere: how is the catalog's corresponding `CatalogCalciteSchema` registered to Calcite's root schema?